File: sdk/python/fluxmeter/event.py
"""Token usage event model."""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field, asdict
from typing import Optional


@dataclass
class TokenEvent:
    """Represents one LLM API call's token usage.

    Supports OpenAI, Anthropic, Google, and custom providers.
    All token fields are optional — set what's available from your provider response.
    """

    customer_id: str
    model_id: str
    provider: str = "openai"

    # Token counts
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    reasoning_tokens: int = 0
    embedding_tokens: int = 0

    # Identity & tracing
    event_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    request_id: Optional[str] = None
    span_id: Optional[str] = None
    parent_span_id: Optional[str] = None  # Links child LLM calls to parent agent run
    session_id: Optional[str] = None

    # Timing
    timestamp: int = field(default_factory=lambda: int(time.time() * 1000))
    latency_ms: int = 0

    # Context
    environment: Optional[str] = None
    metadata: Optional[dict[str, str]] = None

    def to_dict(self) -> dict:
        """Serialize to dict with camelCase keys (matches Java consumer)."""
        d = {
            "eventId": self.event_id,
            "customerId": self.customer_id,
            "provider": self.provider,
            "modelId": self.model_id,
            "inputTokens": self.input_tokens,
            "outputTokens": self.output_tokens,
            "cacheReadTokens": self.cache_read_tokens,
            "cacheWriteTokens": self.cache_write_tokens,
            "reasoningTokens": self.reasoning_tokens,
            "embeddingTokens": self.embedding_tokens,
            "timestamp": self.timestamp,
            "latencyMs": self.latency_ms,
        }
        if self.request_id:
            d["requestId"] = self.request_id
        if self.span_id:
            d["spanId"] = self.span_id
        if self.parent_span_id:
            d["parentSpanId"] = self.parent_span_id
        if self.session_id:
            d["sessionId"] = self.session_id
        if self.environment:
            d["environment"] = self.environment
        if self.metadata:
            d["metadata"] = self.metadata
        return d

    @property
    def total_tokens(self) -> int:
        return (
            self.input_tokens
            + self.output_tokens
            + self.cache_read_tokens
            + self.cache_write_tokens
            + self.reasoning_tokens
            + self.embedding_tokens
        )

File: sdk/js/src/event.ts
/** Token usage event — matches spec/schema/token-event-v1.json */

export interface TokenEvent {
  eventId?: string;
  customerId: string;
  modelId: string;
  provider?: string;
  inputTokens?: number;
  outputTokens?: number;
  cacheReadTokens?: number;
  cacheWriteTokens?: number;
  reasoningTokens?: number;
  embeddingTokens?: number;
  requestId?: string;
  spanId?: string;
  parentSpanId?: string;
  sessionId?: string;
  timestamp?: number;
  latencyMs?: number;
  environment?: string;
  metadata?: Record<string, string>;
}

export function totalTokens(e: TokenEvent): number {
  return (
    (e.inputTokens ?? 0) +
    (e.outputTokens ?? 0) +
    (e.cacheReadTokens ?? 0) +
    (e.cacheWriteTokens ?? 0) +
    (e.reasoningTokens ?? 0) +
    (e.embeddingTokens ?? 0)
  );
}

export function toEventPayload(event: TokenEvent): Record<string, unknown> {
  const d: Record<string, unknown> = {
    customerId: event.customerId,
    modelId: event.modelId,
    provider: event.provider ?? "openai",
    inputTokens: event.inputTokens ?? 0,
    outputTokens: event.outputTokens ?? 0,
    cacheReadTokens: event.cacheReadTokens ?? 0,
    cacheWriteTokens: event.cacheWriteTokens ?? 0,
    reasoningTokens: event.reasoningTokens ?? 0,
    embeddingTokens: event.embeddingTokens ?? 0,
    timestamp: event.timestamp ?? Date.now(),
    latencyMs: event.latencyMs ?? 0,
  };
  if (event.eventId) d.eventId = event.eventId;
  if (event.requestId) d.requestId = event.requestId;
  if (event.spanId) d.spanId = event.spanId;
  if (event.parentSpanId) d.parentSpanId = event.parentSpanId;
  if (event.sessionId) d.sessionId = event.sessionId;
  if (event.environment) d.environment = event.environment;
  if (event.metadata) d.metadata = event.metadata;
  return d;
}
